File: 96-Unique Binary Search Trees-不同的二叉搜索树/96-Unique Binary Search Trees-不同的二叉搜索树.java
class Solution {
    public int numTrees(int n) {
        int[] dp = new int[n + 1];
        dp[0] = 1;
        dp[1] = 1;
        for(int i = 2; i <= n; i++){
            for(int j = 1; j <= i; j++){
                dp[i] += dp[j - 1] * dp[i - j];
            }
        }
        return dp[n];

        // long num = 1;
        // for(int i = 0; i < n; i++){
        //     num = num * 2 * (2 * i + 1) / (i + 2);
        // }
        // return (int)num;
    }
    
}

File: 96-Unique Binary Search Trees-不同的二叉搜索树/96-Unique Binary Search Trees-不同的二叉搜索树.js
/**
 * @param {number} n
 * @return {number}
 */
 var numTrees = function(n) {
    // let nums = [1, 1]

    // for(let i = 2; i <= n; i++){
    //     nums[i] = 0
    //     for(let j = 1; j <= i; j++){
    //         nums[i] += nums[i - j] * nums[j - 1]
    //     }
    // }
    // return nums[n]

    let num = 1
    for(let i = 0; i < n; i++){
        num = num * 2 *(2 * i + 1) / (i + 2)
    }
    return num
};